File: src/filepattern2/filepattern.py
from . import backend
import re, pathlib, typing


class PatternObject:
    def __init__(self, file_pattern, block_size):
        self._file_pattern = file_pattern
        self._block_size = block_size

    def get_matching(self, **kwargs) -> list:
        """Get all filenames matching specific values

        Args:
            **kwargs: One of the variables contained in the pattern

        Returns:
            List of matching files
        """

        mapping = []
        for key, value in kwargs.items():
            mapping.append((key, value))

        if self._block_size == "":
            try:
                return self._file_pattern.getMatching(mapping, False)
            except Exception as e:
                print(e)
        else:
            return self._get_matching_out_of_core(mapping)

    def _get_matching_out_of_core(self, mapping):
        try:
            self._file_pattern.getMatching(mapping)

            while True:
                matching = self._get_matching_block()
                if len(matching) == 0:
                    break

                for match in matching:
                    yield match

        except ValueError as e:
            print(e)

    def _get_matching_block(self) -> list:
        """
        Returns block of mathing files of size less than or equal to block_size.

        Must be called after making a call to get_matching.

        @return list Block of matching files.
        """

        try:
            return self._file_pattern.getMatchingBlock()
        except ValueError as e:
            print(e)

    def get_occurrences(self, mapping):
        """
        Returns the unique values for each variable along with the number of occurrences for each value.

        Args:
            **kwargs: Each keyword argument must be a variable. If no arguments are supplied, the occurrences
            for every variable will be returned.

        Returns:
            Dictionary of variables mapped to values where each value is mapped to the number of occurrences.
        """

        return self._file_pattern.getOccurrences(mapping)

    def get_unique_values(self, vec) -> list:
        """Returns the unique values for each variable.

        This method returns a dictionary of provided variables to a list of all unique occurrences. If no variables are provided,
        all variables will be returned.

        Args:
            **args: Variables to get the occurrences of. All variables will be returned if no arguments are provided.

        Returns:
            Dictionary of variables mapped to values.
        """

        return self._file_pattern.getUniqueValues(vec)

    def output_name(self, files: list = []) -> str:
        """Returns a single filename that captures variables from a list of files.

        Given a list of files, this method will return a single filename that captures the variables from each
        file in the list. If a variable is constant through the list, the variable value will be in the returned
        name. If a variable is not constant, the minimum and maximum values will appear in the returned name in
        the form "(min-max)".

        Args:
            files: List of files to get a single filename of.

        Returns:
            A string that captures the variable values from each file in files.

        """

        return self._file_pattern.outputName(files)
    
    def get_variables(self):
        
        return self._file_pattern.getVariables()

    def __call__(self, group_by: typing.Union[str, list] = [], **kwargs):
        """Iterate through files parsed using a filepattern

        This method returns an iterable of filenames matched to the filepattern. If
        a group_by variable is provided, lists of files where the variable is held constant are
        returned on each call.
        
        Note that the `group_by` argument works in the inverse of the previous version of `filepattern`. 
        The variable passed to `group_by` will be held constant rather than the other variables remaining constant.

        Args:
            group_by: A string consisting of a single variable or a list of variables to group filenames by.
        """
        
        mapping = []
        for key, value in kwargs.items():
            mapping.append((key, value))
            
        self._file_pattern.setMatchingVariables(mapping);
        
        if(group_by == [] or group_by == "" or group_by == [""]):
            return self
        
        if (isinstance(group_by, str)):
            group_by = [group_by]
                
        self._file_pattern.setGroup(group_by)
        if self._block_size == "":  
            if len(group_by) == 0 or group_by[0] != "":
                if(len(mapping) == 0):
                    self._file_pattern.groupBy(group_by, [])
                else:
                    self._file_pattern.groupBy(group_by, self._file_pattern.getMatching(mapping, False))
                    
            return self
        
        if len(group_by) == 0 or (group_by != [""] and len(group_by) != 1):
            self._file_pattern.setGroup(group_by)

        return self

    def _length(self):
        return self._file_pattern.currentBlockLength()

    def __iter__(self):
        """Returns an iterator of files matched to the pattern"""
        if self._block_size == "":
            for file in self._file_pattern.__iter__():
                yield file
        else:
            while True:

                for block in self._file_pattern.__iter__():

                    if self._length() == 0:
                        break

                    yield block

                if self._length() == 0:
                    break
                
    def __getitem__(self, key):
        if(type(key) == int): return self._file_pattern.getItem(key)
        if(type(key) == list): return self._file_pattern.getItemList(key)
        
        slc = [key.start, key.stop, key.step]
        if(slc[0] == None): slc[0] = 'None'
        if(slc[1] == None): slc[1] = 'None'
        if(slc[2] == None): slc[2] = 'None'
        
        return self._file_pattern.getSlice(slc)
    
    def path(self):
        
        return pathlib.Path(self._file_pattern.getPath())


class FilePattern(PatternObject):
    """
    Class to create a FilePattern object.

    This class take in in 4 arguments: path, pattern, block_size, and recursive. For the path,
    either a path to a directory, text file, or stitching vector may be provided. ``filepattern2`` will
    then match the filenames in the directory, or each line of the text file, to the provided ``pattern``.

    The ``block_size`` parameter allows for out of core processing, which consume ``block_size`` amount of memory at most.

    The ``recursive`` parameter enables recursive iteration of subdirectories when a directory is passed as ``path``. In
    this case ``filepattern2`` will iterate over the subdirectories, storing filenames with the same basename in the same
    group.


    Args:
            path: Path to directory or text file
            pattern: Pattern to compare each filename to
            block_size: Maximum amount of RAM to consume at once. Defaults to "".
            recursive: Iterate over subdirectories. Defaults to False.
    """

    def __init__(
        self,
        path: str,
        pattern: str = "",
        block_size: str = "",
        recursive: bool = False,
        suppress_warnings = False
    ):
        """Constructor of the Pattern class. The path argument can either be a directory, a text file,
        or a stitching vector. Passing in the optional argument `block_size` will
        create an ExternalFilePattern object, which will process the directory in blocks which consume less
        than or equal to `block_size` of memory.

        Just the path may be passed in the pattern is contained within the path. In this case,
        the names of the subdirectories are captured if they are named is the same manner as the pattern.
        For example, if just the path 'path/to/files/{channel: c+}/img_r{r:d+}_c{c:d+}.tif' is passed,
        the names of the channel subfolders will be captured for each file.

        Args:
            path: Path to directory or text file
            pattern: Pattern to compare each filename to
            block_size: Maximum amount of RAM to consume at once. Defaults to "".
            recursive: Iterate over subdirectories. Defaults to False.
        """

        
        path = str(path) # change path type to string to support pathlib paths
        
        if path.endswith(".txt"):

            with open(path) as infile:
                line = infile.readline().rstrip()

            if re.match(r"file\: .+?; corr\: .+?; position\: .+?; grid\: .+?;", line):
                if block_size == "":
                    self._file_pattern = backend.InternalVectorPattern(path, pattern, suppress_warnings)
                else:
                    self._file_pattern = backend.ExternalVectorPattern(
                        path, pattern, block_size, suppress_warnings
                    )
                # self._file_pattern = backend.InternalVectorPattern(path, pattern) if block_size == '' \
                #                    else backend.ExternalVectorPattern(path, pattern, block_size=block_size)
            else:
                if block_size == "":
                    self._file_pattern = backend.StringPattern(path, pattern, suppress_warnings)
                else:
                    self._file_pattern = backend.ExternalStringPattern(
                        path, pattern, block_size, suppress_warnings
                    )
                # self._file_pattern = backend.StringPattern(path, pattern) if block_size == '' \
                #                    else backend.ExternalStringPattern(path, pattern, block_size=block_size)
        else:
            if block_size == "":
                self._file_pattern = backend.FilePattern(path, pattern, recursive, suppress_warnings)
            else:
                self._file_pattern = backend.ExternalFilePattern(
                    path, pattern, block_size, recursive, suppress_warnings
                )
            # self._file_pattern = backend.FilePattern(path, pattern, recursive) if block_size == '' \
            #                    else backend.ExternalFilePattern(path, pattern, recursive = recursive, block_size=block_size)

        super().__init__(self._file_pattern, block_size)

    def get_occurrences(self, **kwargs):
        """
        Returns the unique values for each variable along with the number of occurrences for each value.

        Args:
            **kwargs: Each keyword argument must be a variable. If no arguments are supplied, the occurrences
            for every variable will be returned.

        Returns:
            Dictionary of variables mapped to values where each value is mapped to the number of occurrences.
        """

        mapping = []
        for key, value in kwargs.items():
            mapping.append((key, value))

        return super(FilePattern, self).get_occurrences(mapping)

    def get_unique_values(self, *args) -> list:
        """Returns the unique values for each variable.

        This method returns a dictionary of provided variables to a list of all unique occurrences. If no variables are provided,
        all variables will be returned.

        Args:
            **args: Variables to get the occurrences of. All variables will be returned if no arguments are provided.

        Returns:
            Dictionary of variables mapped to values.
        """
        vec = []
        for str in args:
            vec.append(str)

        return super().get_unique_values(vec)

    def output_name(self, files: list = []) -> str:
        """Returns a single filename that captures variables from a list of files.

        Given a list of files, this method will return a single filename that captures the variables from each
        file in the list. If a variable is constant through the list, the variable value will be in the returned
        name. If a variable is not constant, the minimum and maximum values will appear in the returned name in
        the form "(min-max)".

        Args:
            files: List of files to get a single filename of.

        Returns:
            A string that captures the variable values from each file in files.

        """

        return super().output_name(files)
    
    def get_path(self) -> pathlib.Path:
        
        return super().path()

File: src/filepattern2/cpp/interface/filepattern.cpp
#include "filepattern.h"

FilePattern::FilePattern(const std::string& path, const std::string& filePattern, const std::string& block_size, bool recursive, bool suppressWarnings) {

    /*
    PatternInitializer pi;
    pi.block_size = block_size;
    pi.path = path;
    pi.file_pattern = filePattern;
    pi.recursive = recursive;
    pi.suppress_warnings = suppressWarnings;
    */

    FilePatternFactory fpf = FilePatternFactory();

    this->fp_ = std::unique_ptr<PatternObject>(fpf.getObject(path, filePattern, block_size, recursive, suppressWarnings));

    if (block_size != "") {
        this->fp_->external = true;
    } else {
        this->fp_->external = false;
    }

}

std::vector<std::pair<std::vector<std::pair<std::string, Types>> , std::vector<Tuple>>> FilePattern::groupBy(std::vector<std::string>& groups) {
    this->fp_->groupBy(groups);
    return this->fp_->valid_grouped_files_;
}

std::vector<Tuple> FilePattern::getMatching(const std::vector<std::tuple<std::string, std::vector<Types>>>& variables) {
    return this->fp_->getMatching(variables);
    /*
    if (this->fp_->group_.size() == 0) {
        return this->fp_->getMatching(variables);
    } else {
        this->fp_->getMatching(variables);
        return this->fp_->getMatchingBlock();
    }
    */
}

std::vector<Tuple> FilePattern::getMatchingBlock() {
    return this->fp_->getMatchingBlock();
}

void FilePattern::setGroup(std::string& groups){
    std::vector<std::string> group_vec = {groups};
    this->fp_->group_ = group_vec;
}

void FilePattern::setGroup(const std::vector<std::string>& groups) {
    this->fp_->group_ = groups;
}

std::map<std::string, std::map<Types, int>> FilePattern::getOccurrences(const std::vector<std::tuple<std::string, std::vector<Types>>>& mapping) {
    return this->fp_->getOccurrences(mapping);
}

std::map<std::string, std::set<Types>> FilePattern::getUniqueValues(const std::vector<std::string>& mapping) {
    return this->fp_->getUniqueValues(mapping);
}

std::string FilePattern::outputName(std::vector<Tuple>& vec) {
    return this->fp_->outputName(vec);
}

std::vector<std::string> FilePattern::getVariables() {
    return this->fp_->getVariables();
}

void FilePattern::next() {
    this->fp_->next();
}

void FilePattern::nextGroup() {
    this->fp_->nextGroup();
}

int FilePattern::currentBlockLength() {
    return this->fp_->currentBlockLength();
}

/*
int FilePattern::getFiles() {
    if(fp_->external) {

        if(fp_->group_.size() != 0 || (fp_->group_.size() != 0 && fp_->group_[0] != "")) {
            this->nextGroup();
            return fp_->current_group_;
        } else {
            this->next(); 
            return fp_->current_block_;
        }

    } else {

        if(fp_->group_.size() != 0 || (fp_->group_.size() != 0 && fp_->group_[0] != "")){
            return fp_->valid_grouped_files_;
        } 
        else{ 
            return fp_->valid_files_;
        }
}
*/

void FilePattern::getNewNaming(std::string& pattern, bool suppressWarnings) {
    this->fp_->getNewNaming(pattern, suppressWarnings);
}

std::vector<Tuple> FilePattern::getSlice(std::vector<Types>& key) {
    return this->fp_->getSlice(key);
}

//std::string FilePattern::swSearch(std::string& pattern, std::string& filename, const std::string& variables) {
//    return this->fp_->swSearch(pattern, filename, variables);
//}


std::string FilePattern::inferPattern(const std::string& path, std::string& variables, const std::string& block_size){

    FilePatternFactory fpf = FilePatternFactory();

    // create dummy object to avoid the need for static methods in virtual class
    std::unique_ptr<PatternObject> fp;
    if (block_size == "") {
        fp = std::unique_ptr<PatternObject>(fpf.getObject(path, "pattern{r:d}", block_size, false, true));
    } else {

        fp = std::unique_ptr<PatternObject>(fpf.getObject(path, "", block_size, false, true));
    }


    return fp->inferPattern(path, variables, block_size);
}

std::string FilePattern::inferPattern(std::vector<std::string>& vec, std::string& variables) {

    FilePatternFactory fpf = FilePatternFactory();
    
    std::unique_ptr<PatternObject> fp = std::unique_ptr<PatternObject>(fpf.getObject(".", "dummy_pattern", "", false, true));

    return fp->inferPattern(vec, variables);
}

Tuple FilePattern::getItem(int key) {
    return this->fp_->getItem(key);
}

std::vector<Tuple> FilePattern::getItemList(std::vector<int>& key) {
    return this->fp_->getItemList(key);
}